### ISSUE TITLE
problems with encoding to Vec<u8>

### ISSUE THREAD

**Issue by godfryd** [3Hren/msgpack-rust#14]
Hello,
I tried to encode to growable buffer (not fixed length one like it is used in example).
I change the example to:

```rust
extern crate rmp as msgpack;
extern crate rustc_serialize;
use rustc_serialize::Encodable;
use msgpack::Encoder;
fn main() {
    let val = (42u8, "the Answer");
    //let mut buf = [0u8; 23];
    let mut buf: Vec<u8> = Vec::new();
    val.encode(&mut Encoder::new(&mut &mut buf[..])).unwrap();
}
```

It panics on val.encode:

    thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Unimplemented', src/libcore/result.rs:731
    stack backtrace:
       1:     0x7fe7d8338e8e - sys::backtrace::write::hd71fc66fa1c984beLqs
       2:     0x7fe7d833be35 - panicking::on_panic::hc1f1f7a7d3f1cd2al8w
       3:     0x7fe7d83352ae - rt::unwind::begin_unwind_inner::h26926b64783ab69bZNw
       4:     0x7fe7d83354fc - rt::unwind::begin_unwind_fmt::ha28396e9e1ce1f5e5Mw
       5:     0x7fe7d833b826 - rust_begin_unwind
       6:     0x7fe7d8365414 - panicking::panic_fmt::h44955cbbc9e564ec19B
       7:     0x7fe7d8328580 - result::Result<T, E>::unwrap::h3504239604260474918
                            at src/libcore/macros.rs:28
       8:     0x7fe7d832804d - main::habc3974e69a85df6iaa
                            at src/lib.rs:14
       9:     0x7fe7d83402a8 - rust_try_inner
      10:     0x7fe7d8340295 - rust_try
      11:     0x7fe7d833d968 - rt::lang_start::h9a2154723c054292i3w
      12:     0x7fe7d832eb9b - main
      13:     0x7fe7d7512a3f - __libc_start_main
      14:     0x7fe7d8327e28 - _start
      15:                0x0 - <unknown>

**Comment by 3Hren** [comment]
Hi!

That's why you explicicly pass the `Vec` as mutable buffer.

The `Encoder` struct accepts `&mut Write` as a writable entity. In your example it will be `&mut [u8]` slice, that can't be auto resized. To work with **exactly** `Vec` just remove slice operator (`&mut [..]`) like this:

```rust
extern crate rmp as msgpack;
extern crate rustc_serialize;

use rustc_serialize::Encodable;
use msgpack::Encoder;

fn main() {
    let val = (42u8, "the Answer");

    let mut buf: Vec<u8> = Vec::new();

    val.encode(&mut Encoder::new(&mut buf)).unwrap();

    assert_eq!(vec![0x92, 0x2a, 0xaa, 0x74, 0x68, 0x65, 0x20, 0x41, 0x6e, 0x73, 0x77, 0x65, 0x72], buf);
}
```

**Comment by godfryd** [comment]
Thanks. I changed the code according to your tip and it works.

Still, shouldn't there be something more specific error than unimplemented?
Or in other words, the real error info is buried so deeply that now it is hard to diagnose what is wrong.

**Comment by 3Hren** [comment]
Agree :)

Already fixed in 1fa3ab1bc5ad7923a45904e6aaee00a9506ea3ef
